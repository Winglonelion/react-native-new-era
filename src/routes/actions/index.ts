import { isNavigationReady, navigationRef } from 'app/Navigation.ref';
import { DrawerActions } from '@react-navigation/native';

export function navigate(name: string, params: Object) {
  if (isNavigationReady() && navigationRef.current) {
    // Perform navigation if the app has mounted
    navigationRef.current.navigate(name, params);
  } else {
    // You can decide what to do if the app hasn't mounted
    // You can ignore this, or add these actions to a queue you can call later
  }
}

export function toggleDrawer() {
  if (isNavigationReady() && navigationRef.current) {
    // Perform navigation if the app has mounted
    navigationRef.current.dispatch(DrawerActions.toggleDrawer());
  } else {
    // You can decide what to do if the app hasn't mounted
    // You can ignore this, or add these actions to a queue you can call later
  }
}
