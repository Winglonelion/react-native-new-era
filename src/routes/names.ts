export namespace ROUTES_TYPES {
  export type HOME = 'HOME';
}

const ROUTES = {
  // stack
  MAIN_STACK: 'MAIN_STACK',
  MODAL_STACK: 'MODAL_STACK',
  AUTHORIZED_STACK: 'AUTHORIZED_STACK',
  MAIN_DRAWER: 'MAIN_DRAWER',
  UNAUTHORIZED_STACK: 'UNAUTHORIZED_STACK',
  // HOME
  HOME: 'HOME',
  NOTIFICATION: 'NOTIFICATION',
  BASE: 'BASE',
  PROFILE: 'PROFILE',
  LOGIN_SCREEN: 'LOGIN_SCREEN',
  REGISTER: 'REGISTER',
  PROMO: 'PROMO',
  SPLASH_SCREEN: 'SPLASH_SCREEN',
  ONBOARDING_SCREEN: 'ONBOARDING_SCREEN',
  SETTING_SCREEN: 'SETTING_SCREEN',
  FEEDBACK_SCREEN: 'FEEDBACK_SCREEN',
};

export const ROUTE_TABS = {
  MAIN_TAB: 'MAIN_TAB',
};

export default ROUTES;
