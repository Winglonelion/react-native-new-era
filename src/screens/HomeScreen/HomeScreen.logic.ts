import { useCallback, useEffect } from 'react';
import homeScreenStore from './HomeScreen.store';

const useHomeScreenLogic = () => {
  const reloadHomeData = useCallback(() => {
    homeScreenStore.fetchPaidHistoryOverview();
    homeScreenStore.fetchPaidOverview();
  }, []);

  useEffect(() => {
    reloadHomeData();
  }, [reloadHomeData]);

  return { reloadHomeData };
};

export default useHomeScreenLogic;
