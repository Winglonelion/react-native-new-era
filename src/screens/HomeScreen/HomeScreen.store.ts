import { computed, flow, makeObservable, observable } from 'mobx';
import { AxiosResponse } from 'axios';
import {
  GetPaidHistoryOverViewResponse,
  GetPaidOverViewResponse,
} from 'api/paid/paid.api.types';
import {
  getPaidHistoryOverViewAPI,
  getPaidOverViewAPI,
} from 'api/paid/paid.api';

class HomeScreenStore {
  paidOverViewData: GetPaidOverViewResponse = {
    gross: NaN,
    regular: NaN,
    over_time: NaN,
    hours: NaN,
    rate: NaN,
    take_home: NaN,
    detail: [
      { type: 'taxes', value: NaN },
      { type: 'benefits', value: NaN },
      { type: 'retirements', value: NaN },
      { type: 'other', value: NaN },
      { type: 'taxes', value: NaN },
      { type: 'taxes', value: NaN },
    ],
  };
  loadingPaidOverView: boolean = false;

  paidHistoryOverViewData: GetPaidHistoryOverViewResponse = [];
  loadingPaidHistoryOverView: boolean = false;

  constructor() {
    makeObservable(this, {
      paidOverViewData: observable,
      paidHistoryOverViewData: observable,
      loadingPaidOverView: observable,
      loadingPaidHistoryOverView: observable,
      homeLoading: computed,
      fetchPaidOverview: flow,
      fetchPaidHistoryOverview: flow,
    });
  }

  get homeLoading() {
    return this.loadingPaidOverView || this.loadingPaidHistoryOverView;
  }

  *fetchPaidOverview() {
    try {
      if (this.loadingPaidOverView) return;
      this.loadingPaidOverView = true;
      const response: AxiosResponse<GetPaidOverViewResponse> =
        yield getPaidOverViewAPI({ userId: '123' });
      this.paidOverViewData = response.data;
      this.loadingPaidOverView = false;
    } catch (error) {
      this.loadingPaidOverView = false;
      console.warn('fetchPaidOverview', { error });
    }
  }

  *fetchPaidHistoryOverview() {
    try {
      if (this.loadingPaidHistoryOverView) return;
      this.loadingPaidHistoryOverView = true;
      const response: AxiosResponse<GetPaidHistoryOverViewResponse> =
        yield getPaidHistoryOverViewAPI({ userId: '123' });
      this.paidHistoryOverViewData = response.data;
      this.loadingPaidHistoryOverView = false;
    } catch (error) {
      this.loadingPaidHistoryOverView = false;
      console.warn('fetchPaidHistoryOverview', { error });
    }
  }
}

/**
 * @instance store session info of a user.
 */
const homeScreenStore = new HomeScreenStore();
export default homeScreenStore;
