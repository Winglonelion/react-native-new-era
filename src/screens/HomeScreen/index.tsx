export { default as default } from './HomeScreen';
