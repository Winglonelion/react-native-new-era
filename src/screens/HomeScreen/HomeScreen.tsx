import React, { useCallback } from 'react';
import { RefreshControl, ScrollView, View } from 'react-native';
import { observer } from 'mobx-react';
import { TouchableWithoutFeedback } from 'react-native-gesture-handler';

import ROUTES from 'routes/names';
import { navigateTo } from 'routes/actions';
import CommonStyles from 'theme/CommonStyles';
import DateTitle from 'components/Common/DateTitle';
import PayHistoryOverView from 'components/Business/PayHistory';

import PaidOverview from './components/PaidOverview';
import homeScreenStore from './HomeScreen.store';
import useHomeScreenLogic from './HomeScreen.logic';

const HomeScreen: React.FC = () => {
  const { reloadHomeData } = useHomeScreenLogic();

  const onPressPaidOverview = useCallback(() => {
    navigateTo(ROUTES.PAID_DETAIL_SCREEN, {
      data: homeScreenStore.paidOverViewData,
    });
  }, []);

  return (
    <View style={CommonStyles.container}>
      <ScrollView
        bounces={false}
        showsVerticalScrollIndicator={false}
        refreshControl={
          <RefreshControl
            refreshing={homeScreenStore.homeLoading}
            onRefresh={reloadHomeData}
          />
        }>
        <DateTitle />
        <TouchableWithoutFeedback
          disabled={homeScreenStore.loadingPaidOverView}
          onPress={onPressPaidOverview}>
          <PaidOverview
            overviewData={homeScreenStore.paidOverViewData}
            loading={homeScreenStore.homeLoading}
          />
        </TouchableWithoutFeedback>
        <PayHistoryOverView
          overviewData={homeScreenStore.paidHistoryOverViewData}
          loading={homeScreenStore.homeLoading}
        />
      </ScrollView>
    </View>
  );
};

export default observer(HomeScreen);
