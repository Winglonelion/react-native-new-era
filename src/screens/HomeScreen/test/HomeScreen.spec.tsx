import React from 'react';
import { shallow } from 'enzyme';
import HomeScreen from '../HomeScreen';
import paidMock from 'mock/networking/paid.mock';

jest.mock('routes/actions', () => ({
  __esModule: true,
  navigateTo: jest.fn(),
}));

import { navigateTo } from 'routes/actions';

paidMock();

function setup() {
  const wrapper = shallow(<HomeScreen />);
  return {
    wrapper,
  };
}

describe('HomeScreen check snapshot', function () {
  test('should match snapshot', function () {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });
});

describe('HomeScreen check events', function () {
  beforeEach(() => {
    const useCallbackSpy = jest.spyOn(React, 'useCallback');
    const useEffectSpy = jest.spyOn(React, 'useEffect');
    useCallbackSpy.mockImplementation(
      (callback: (...args: any[]) => any, _: any) => {
        return callback();
      },
    );
    useEffectSpy.mockImplementationOnce(
      (callback: (...args: any[]) => any, _: any) => {
        return callback();
      },
    );
  });

  test('should click overview work', function () {
    const { wrapper } = setup();
    wrapper.find('ForwardRef').simulate('onPress');
    expect(navigateTo).toBeCalledTimes(1);
  });
});
