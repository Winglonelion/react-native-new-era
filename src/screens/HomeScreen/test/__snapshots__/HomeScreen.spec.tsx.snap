// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen check snapshot should match snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
>
  <ScrollView
    bounces={false}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    showsVerticalScrollIndicator={false}
  >
    <Memo() />
    <ForwardRef
      delayLongPress={600}
      disabled={false}
      extraButtonProps={
        Object {
          "rippleColor": "transparent",
        }
      }
      onPress={[Function]}
    >
      <PaidOverview
        loading={false}
        overviewData={
          Object {
            "detail": Array [
              Object {
                "type": "taxes",
                "value": NaN,
              },
              Object {
                "type": "benefits",
                "value": NaN,
              },
              Object {
                "type": "retirements",
                "value": NaN,
              },
              Object {
                "type": "other",
                "value": NaN,
              },
              Object {
                "type": "taxes",
                "value": NaN,
              },
              Object {
                "type": "taxes",
                "value": NaN,
              },
            ],
            "gross": NaN,
            "hours": NaN,
            "over_time": NaN,
            "rate": NaN,
            "regular": NaN,
            "take_home": NaN,
          }
        }
      />
    </ForwardRef>
    <Memo()
      loading={false}
      overviewData={Array []}
    />
  </ScrollView>
</View>
`;
